net: if: Add IPv4 support to multicast monitor

mrfuchs opened the pull request:
Make multicast group join/leave monitor support both IPv6 and IPv4 addresses.

Fixes #26585

Signed-off-by: Markus Fuchs <markus.fuchs@ch.sauter-bc.com>

mrfuchs committed:
net: if: Add IPv4 support to multicast monitor

Make multicast group join/leave monitor support both IPv6 and IPv4
addresses.

Fixes #26585

Signed-off-by: Markus Fuchs <markus.fuchs@ch.sauter-bc.com>

jukkar commented on subsys/net/ip/ipv6_mld.c:
```diff
@@ -201,8 +201,6 @@ int net_ipv6_mld_join(struct net_if *iface, const struct in6_addr *addr)
 
 	net_if_ipv6_maddr_join(maddr);
 
-	net_if_mcast_monitor(iface, addr, true);
```
Why this change? I think we should start to monitor multicast addresses only when the user wants to join the multicast group, and not when a multicast address is added to the network interface.

mrfuchs replied:
I wanted the `net_if_ipv*_maddr_add()` and `net_if_ipv*_maddr_rm()` functions to behave the same way for IPv4 and IPv6 as the mcast monitor is going to be used by Ethernet drivers to adjust their multicast address filters.
However, you are right, the existing behavior should be maintained (not only because it's the _multicast group join/leave monitor_). I reverted this change.

jukkar replied:
Note that we currently support also IPv4 IGMP (so similar API as for IPv6 MLD). There exists `net_ipv4_igmp_join()` and `net_ipv4_igmp_leave()` functions are you should tweak those for IPv4 support.

mrfuchs replied:
These functions make use of `net_if_ipv4_maddr_add()` and `net_if_ipv4_maddr_rm()` which now call `net_if_mcast_monitor()`. How should these functions be tweaked?

https://github.com/zephyrproject-rtos/zephyr/blob/8dcfa913d0f2d0675b4427bbc7d14302a9f13903/subsys/net/ip/igmp.c#L264-L275
https://github.com/zephyrproject-rtos/zephyr/blob/8dcfa913d0f2d0675b4427bbc7d14302a9f13903/subsys/net/ip/igmp.c#L294-L304

jukkar replied:
You would need to remove the code you added to `net_if_ipv4_maddr_add/rm` and just have something like this
```
diff --git a/subsys/net/ip/igmp.c b/subsys/net/ip/igmp.c
index a36444acab..0398f324ae 100644
--- a/subsys/net/ip/igmp.c
+++ b/subsys/net/ip/igmp.c
@@ -285,6 +285,8 @@ int net_ipv4_igmp_join(struct net_if *iface, const struct in_addr *addr)
 
        net_if_ipv4_maddr_join(maddr);
 
+       net_if_mcast_monitor(iface, addr, true);
+
        net_mgmt_event_notify_with_info(NET_EVENT_IPV4_MCAST_JOIN, iface,
                                        &maddr->address.in_addr,
                                        sizeof(struct in_addr));
@@ -312,6 +314,8 @@ int net_ipv4_igmp_leave(struct net_if *iface, const struct in_addr *addr)
 
        net_if_ipv4_maddr_leave(maddr);
 
+       net_if_mcast_monitor(iface, addr, false);
+
        net_mgmt_event_notify_with_info(NET_EVENT_IPV4_MCAST_LEAVE, iface,
                                        &maddr->address.in_addr,
                                        sizeof(struct in_addr));
```

mrfuchs replied:
Got it, but with these changes my patch would implement something different than I had in mind. My intention was to allow Ethernet drivers to fit their multicast filters to the actual requirements (dynamically and at runtime). At the moment, they are either using Kconfig like eth_mcux.c...

https://github.com/zephyrproject-rtos/zephyr/blob/417647f05715cc183df43ffd841e895c03a486ad/drivers/ethernet/eth_mcux.c#L915-L918
https://github.com/zephyrproject-rtos/zephyr/blob/417647f05715cc183df43ffd841e895c03a486ad/drivers/ethernet/eth_mcux.c#L973-L975

... or - even worse - disable the multicast filter completely like eth_stm32_hal.c (cf. ad8f837):

https://github.com/zephyrproject-rtos/zephyr/blob/417647f05715cc183df43ffd841e895c03a486ad/drivers/ethernet/eth_stm32_hal.c#L135-L147

Staying at mDNS as an example, removing my code from `net_if_ipv4_maddr_add()` would make it impossible for the Ethernet driver to use the mcast monitor to adjust its multicast address filter as the mDNS module is calling `net_if_ipv4_maddr_add()` (in case of IPv4 and `net_ipv6_mld_join()` in case of IPv6 - thus my first approach):

https://github.com/zephyrproject-rtos/zephyr/blob/417647f05715cc183df43ffd841e895c03a486ad/subsys/net/lib/dns/mdns_responder.c#L537-L547

As a STM32 and Ethernet stack user struggling with the heavy load that the disabled multicast filters bring, I might better follow the design of the eth_mcux.c driver...

jukkar replied:
For the multicast to work properly in all cases, the application needs to join the multicast group, it is not enough just to add the multicast address to the network interface. For mDNS, this probably does not matter too much but it does not harm to join the group in that case too. So it might be needed for the mDNS library to join the group instead of just adding the multicast address to the interface. This would require some code changes in mDNS to make it work properly.

mrfuchs replied:
> You would need to remove the code you added to `net_if_ipv4_maddr_add/rm` and just have something like this [...]

Followed your advice. Thanks.

mrfuchs replied:
> So it might be needed for the mDNS library to join the group instead of just adding the multicast address to the interface. This would require some code changes in mDNS to make it work properly.

I changed that in an additional commit.

jukkar reviewed (commented):
The `drivers/can/can_net.c` is also using the multicast monitor API, you need to change that too.

rlubos commented on drivers/ethernet/eth_mcux.c:
```diff
@@ -1029,14 +1029,18 @@ static int eth_init(const struct device *dev)
 	struct net_eth_addr mac_addr;
 
-	net_eth_ipv6_mcast_to_mac_addr(addr, &mac_addr);
+	if (addr->family != AF_INET6) {
```
I'd be good to move the Ethernet driver changes to a separate commit, they don't seem to be cover in the commit message (and are not related to the core IP stack changes).

mrfuchs replied:
Done. Thanks!

mrfuchs committed:
drivers: Update drivers to use new multicast monitor API

Update drivers regarding latest multicast group join/leave monitor
changes which now supports both IPv6 and IPv4 multicast addresses.

Signed-off-by: Markus Fuchs <markus.fuchs@ch.sauter-bc.com>

mrfuchs commented:
> The `drivers/can/can_net.c` is also using the multicast monitor API, you need to change that too.

I somehow missed that. Thanks for the hint.

alexanderwachter reviewed (approved):
CAN part LGTM

jukkar reviewed (changes requested):
This is an API change (although very minor and probably does not affect any component outside current drivers), so we should document this in release note.

hakehuang commented on drivers/ethernet/eth_mcux.c:
```diff
@@ -1029,14 +1029,18 @@ static int eth_init(const struct device *dev)
 
 #if defined(CONFIG_NET_IPV6)
```
looks like in driver level this feature is not enabled for ipv4. see below, and above macro is limited to ipv6
```
static void eth_iface_init(struct net_if *iface)
{
	const struct device *dev = net_if_get_device(iface);
	struct eth_context *context = dev->data;

#if defined(CONFIG_NET_IPV6)
	static struct net_if_mcast_monitor mon;

	net_if_mcast_mon_register(&mon, iface, net_if_mcast_cb);
#endif /* CONFIG_NET_IPV6 */
```

mrfuchs replied:
As for the drivers that are currently making use of the mcast monitor, I just tried to make sure that their functionality is preserved.

hakehuang reviewed (commented):
seems not enable ipv4 multicast in driver level, do we plan to do it in another PR?

mrfuchs commented:
> seems not enable ipv4 multicast in driver level, do we plan to do it in another PR?

Yes, that should be done in other PRs. I'll take care of the STM32 Ethernet driver as that's the hardware I have at hand.

> looks like in driver level this feature is not enabled for ipv4. see below, and above macro is limited to ipv6

As for the drivers that are currently making use of the mcast monitor, I just tried to make sure that their functionality is preserved.

mrfuchs committed:
net: dns: Make mdns and llmnr responders join their multicast groups

As with IPv6, the mdns and llmnr responders should join their multicast
groups for IPv4 instead of just adding the multicast address to the
interface.

Signed-off-by: Markus Fuchs <markus.fuchs@ch.sauter-bc.com>

mrfuchs committed:
doc: release: Add note about multicast monitor's new IPv4 support

Add release note about multicast group join/leave monitor API change.

Signed-off-by: Markus Fuchs <markus.fuchs@ch.sauter-bc.com>

mrfuchs commented:
> This is an API change (although very minor and probably does not affect any component outside current drivers), so we should document this in release note.

@jukkar, added a short description to the release notes.

jukkar reviewed (approved):
LGTM

cfriedt commented on doc/releases/release-notes-2.7.rst:
```diff
@@ -117,6 +117,12 @@ Stable API changes in this release
     structure contains two members: ``handles``, which was moved from
     :c:struct:`bt_gatt_read_params`, and ``variable``.
 
+* Networking
```
since this has been converted from a bug to an enhancement, does that mean it can no longer go in for v2.7.0 @nashif ?
